File: algo/40. Combination Sum II.js
// @9 DFS

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    // 1. definition
    var helper = function(res, collection, index, remain_target) {
        if (index > candidates.length) {
            return;
        }
        
        var collection = Array.from(collection);
        // 3. exit
        if (remain_target == 0) {
            res.push(collection);
        }
        
        // 2. split
        for (var i = index; i < candidates.length; i++) {
            if (candidates[i] > remain_target) {
                // note that it should be continue here instead of break
                continue;
            }
            
            if (i!=index && candidates[i] === candidates[i-1]) {
                continue;
            }
            
            collection.push(candidates[i]);
            helper(res, collection, i+1, remain_target - candidates[i]);
            collection.pop();
        }
    }
    
    var res = [];
    if (candidates.length == 0) {
        return res;
    }
    
    candidates.sort();
    helper(res, [], 0, target);
    return res;
};

File: round3/40. Combination Sum II.cs
public class Solution {
    public IList<IList<int>> CombinationSum2(int[] candidates, int target) {
        Array.Sort(candidates);
        
        IList<IList<int>> res = new List<IList<int>>();
        DFSHelper(candidates, target, 0, new List<int>(), res);
        return res;
    }
    
    public void DFSHelper(int[] candidates, int target, int index, IList<int> currRes, IList<IList<int>> res) {
        if (target == 0) {
            res.Add(new List<int>(currRes));
            return;
        }
        
        if (target < 0) {
            return;
        }
        
        for (int i=index; i<candidates.Length; i++) {
            if (i != index && candidates[i] == candidates[i-1]) continue;
            
            currRes.Add(candidates[i]);
            DFSHelper(candidates, target-candidates[i], i+1, currRes, res);
            currRes.RemoveAt(currRes.Count-1);
        }
    }
}